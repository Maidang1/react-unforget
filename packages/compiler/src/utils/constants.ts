export const RUNTIME_MODULE_NAME = "@react-unforget/runtime";
export const RUNTIME_MODULE_CREATE_CACHE_HOOK_NAME = "useCreateCache$unforget";
export const DEFAULT_CACHE_VARIABLE_NAME = "$unforgetCache";
export const DEFAULT_CACHE_COMMIT_VARIABLE_NAME = "$unforgetCommit";

export const DEFAULT_UNUSED_VARIABLE_NAME = "_unused";
export const DEFAULT_UNWRAPPED_VARIABLE_NAME = "_unwrapped";

export const RUNTIME_MODULE_CACHE_ENQUEUE_METHOD_NAME = "e";
export const RUNTIME_MODULE_CACHE_VALUE_PROP_NAME = "v";
export const RUNTIME_MODULE_CACHE_IS_NOT_SET_PROP_NAME = "n";
