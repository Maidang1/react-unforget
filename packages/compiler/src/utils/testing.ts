import * as standalone from "@babel/standalone";
import * as babel from "@babel/core";
import * as generateBase from "@babel/generator";
import traverse from "@babel/traverse";
import type * as t from "@babel/types";
import * as fs from "fs";
import * as path from "path";
import { visitProgram } from "~/visit-program";
// @ts-expect-error The module has no types
import babelJsxPlugin from "@babel/plugin-syntax-jsx";

export function transformWithStandalone(input: string) {
  const { code } = standalone.transform(input, {
    configFile: false,
    plugins: [
      babelJsxPlugin,
      {
        visitor: {
          Program: (path: babel.NodePath<t.Program>) => {
            visitProgram(path);
            path.skip();
          },
        },
      },
    ],
  })!;
  return code;
}

export function transformWithCore(input: string) {
  const { code } = babel.transform(input, {
    configFile: false,
    plugins: [
      babelJsxPlugin,
      {
        visitor: {
          Program: (path: babel.NodePath<t.Program>) => {
            visitProgram(path);
          },
        },
      },
    ],
  })!;
  return code;
}

export function transformWithParseAndCast(input: string) {
  const root = parse(input);

  visitProgram(root);

  return String(root);
}

export function parse(input: string, extraPlugins: babel.PluginItem[] = []) {
  const ast = babel.parse(input, {
    configFile: false,
    plugins: ["@babel/plugin-syntax-jsx", ...extraPlugins],
  })!;

  let result: babel.NodePath<t.Program> | null = null;
  traverse(ast, {
    enter(path) {
      result = path as babel.NodePath<t.Program>;
      path.skip();
    },
  });

  return result!;
}

export function loadFixture(fixturePath: string) {
  const fixture = path.resolve(__dirname, "../tests/fixtures", fixturePath + ".jsx");

  const fileContent = fs.readFileSync(fixture, "utf-8");

  return fileContent;
}

export function parseFixture(fixturePath: string) {
  return parse(loadFixture(fixturePath));
}

export function generate(path: babel.types.Node) {
  return generateBase.default(path).code;
}
