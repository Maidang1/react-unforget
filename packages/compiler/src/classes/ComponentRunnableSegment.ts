import * as babel from "@babel/core";
import { Binding } from "@babel/traverse";
import { convertStatementToSegmentCallable } from "~/ast-factories/convert-statement-to-segment-callable";
import { getBlockStatementsOfPath } from "~/utils/get-block-statements-of-path";
import { getReferencedVariablesInside } from "~/utils/get-referenced-variables-inside";
import { Component } from "./Component";
import {
  ComponentMutableSegment,
  SegmentTransformationResult,
} from "./ComponentMutableSegment";
import { reorderByTopology } from "~/utils/reorder-by-topology";

export class ComponentRunnableSegment extends ComponentMutableSegment {
  private mapOfReturnStatementToReferencedBindings = new Map<
    babel.NodePath<babel.types.ReturnStatement>,
    Binding[]
  >();

  private blockReturnStatement: babel.NodePath<babel.types.ReturnStatement> | null =
    null;

  constructor(
    component: Component,
    parent: ComponentMutableSegment | null = null,
    private statement: babel.NodePath<babel.types.Statement>
  ) {
    super(component, parent, "ComponentRunnableSegment");
  }

  isRoot() {
    return this.parent === null;
  }

  get hasReturnStatement() {
    return !!this.blockReturnStatement;
  }

  computeDependencyGraph() {
    this.dependencies.clear();

    const path = this.path;

    const blockStatementChildren = getBlockStatementsOfPath(path);

    blockStatementChildren.forEach((blockStatement) => {
      this.component.addRunnableSegment(blockStatement);
    });

    blockStatementChildren.forEach((currentPath) => {
      const statements = currentPath.get("body");

      if (path.isBlockStatement() && currentPath === path) {
        statements.forEach((statement) => {
          const returnStatement = statement;
          if (this.hasReturnStatement) {
            return;
          }

          if (returnStatement.isReturnStatement()) {
            this.blockReturnStatement = returnStatement;

            const bindings: Binding[] = [];
            getReferencedVariablesInside(returnStatement).forEach((binding) => {
              bindings.push(binding);

              this.component.addComponentVariable(binding);
            });

            this.mapOfReturnStatementToReferencedBindings.set(
              returnStatement,
              bindings
            );
          } else {
            this.component.addRunnableSegment(statement);
          }
        });
      } else {
        const child = this.component.addRunnableSegment(currentPath);
        this.children.add(child);
      }
    });
  }

  applyTransformation(performReplacement = true) {
    const path = this.path;

    const transformationsToPerform: (() => babel.NodePath[] | null)[] = [];
    const callables: babel.types.Statement[] = [];

    if (path.isBlockStatement()) {
      const statements = path.get("body");

      const segmentsMap = this.component.getStatementsToMutableSegmentMap();

      const reorderedStatements = reorderByTopology(statements, segmentsMap);

      reorderedStatements.forEach((statement) => {
        const segment = segmentsMap.get(statement);

        const transformation = segment?.applyTransformation();
        if (transformation) {
          const callStatement = this.makeSegmentCallStatement(transformation);

          transformationsToPerform.push(() =>
            transformation.prformTransformation()
          );

          if (callStatement) {
            callables.push(...callStatement);
          }
        } else {
          callables.push(statement.node);
        }
      });

      this.children.forEach((child) => {
        child.applyTransformation();
      });

      const newNodes = transformationsToPerform
        .flatMap((transformation) => transformation()?.map(({ node }) => node))
        .filter((v): v is babel.types.Statement => v !== null)
        .concat(callables);

      path.node.body = newNodes;

      return null;
    }

    const anyChildrenHasReturnStatement = Array.from(this.children).some(
      (child) => child.isComponentRunnableSegment() && child.hasReturnStatement
    );

    const hasHookCall = this.hasHookCall();

    const dependencyConditions = this.makeDependencyCondition();

    const { prformTransformation, segmentCallableId, replacements } =
      convertStatementToSegmentCallable(path, {
        performReplacement,
        segmentCallableId: this.getSegmentCallableId(),
      });

    return {
      dependencyConditions,
      hasHookCall,
      prformTransformation,
      segmentCallableId,
      replacements,
      hasReturnStatement: anyChildrenHasReturnStatement,
    } satisfies SegmentTransformationResult;
  }

  getParentStatement() {
    return this.path;
  }

  getBlockStatements() {
    return getBlockStatementsOfPath(this.path);
  }

  get path() {
    return this.statement;
  }
}
