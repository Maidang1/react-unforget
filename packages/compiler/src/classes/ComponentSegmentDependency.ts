import * as t from "@babel/types";
import { ComponentVariable } from "./ComponentVariable";

type AccessorNode =
  | t.MemberExpression
  | t.OptionalMemberExpression
  | t.Identifier
  | t.PrivateName;

export class AccessChainItem {
  public nextComputed = false;
  public right?: AccessChainItem = undefined;

  constructor(
    public id: string,
    public idExpression: t.Expression
  ) {}

  toString() {
    let currentString = this.id;
    if (this.right) {
      currentString += this.nextComputed
        ? `[${this.right.toString()}]`
        : `.${this.right.toString()}`;
    }
    return currentString;
  }
}

function isAccessorNode(node: t.Node): node is AccessorNode {
  return (
    t.isMemberExpression(node) ||
    t.isOptionalMemberExpression(node) ||
    t.isIdentifier(node) ||
    t.isPrivateName(node)
  );
}

// When the variable is used in a member expression, we should optimize comparisons to the last member of member expression as well
export class ComponentSegmentDependency {
  private root: AccessChainItem;

  constructor(
    public componentVariable: ComponentVariable,
    public accessorNode: AccessorNode
  ) {
    let currentAccessChainItem: AccessChainItem | null = null;

    // Start from the root of member expression
    let currentAccessorNode: AccessorNode | null = accessorNode;

    do {
      let newAccessChainItem: AccessChainItem | null = null;

      if (
        t.isMemberExpression(currentAccessorNode) ||
        t.isOptionalMemberExpression(currentAccessorNode)
      ) {
        if (currentAccessChainItem) {
          currentAccessChainItem.nextComputed = currentAccessorNode.computed;
        }

        const property = currentAccessorNode.property;
        let name = "";

        if (t.isIdentifier(property)) {
          name = property.name;
        } else if (t.isStringLiteral(property)) {
          name = property.value;
        } else if (t.isNumericLiteral(property)) {
          name = property.value.toString();
        } else if (t.isPrivateName(property)) {
          name = "#" + property.id.name;
        } else if (t.isBigIntLiteral(property)) {
          name = property.value + "n";
        } else if (
          t.isTemplateLiteral(property) &&
          property.expressions.length === 0 &&
          property.quasis.length === 1
        ) {
          name = property.quasis[0]!.value.raw;
        }

        if (name) {
          newAccessChainItem = new AccessChainItem(name, currentAccessorNode);

          if (isAccessorNode(currentAccessorNode.object)) {
            currentAccessorNode = currentAccessorNode.object;
          } else {
            currentAccessorNode = null;
          }
        } else {
          currentAccessorNode = null;
        }
      } else if (t.isIdentifier(currentAccessorNode)) {
        newAccessChainItem = new AccessChainItem(
          currentAccessorNode.name,
          currentAccessorNode
        );

        currentAccessorNode = null;
      }

      if (newAccessChainItem) {
        if (currentAccessChainItem) {
          newAccessChainItem.right = currentAccessChainItem;
        }
      }

      currentAccessChainItem = newAccessChainItem;
    } while (currentAccessorNode);

    this.root = currentAccessChainItem!;
  }

  getRootDependencyLink() {
    return this.root;
  }

  stringify() {
    return this.root.toString();
  }

  equals(other: ComponentSegmentDependency) {
    return (
      this.componentVariable === other.componentVariable &&
      this.stringify() === other.stringify()
    );
  }

  getMemberExpression(replacementId: t.Expression) {
    let endOfChaing: AccessChainItem | null = this.root;

    while (endOfChaing.right) {
      endOfChaing = endOfChaing.right;
    }

    const endOfChainExpression = t.cloneNode(endOfChaing.idExpression);

    if (
      t.isMemberExpression(endOfChainExpression) ||
      t.isOptionalMemberExpression(endOfChainExpression)
    ) {
      endOfChainExpression.object = replacementId;
      return endOfChainExpression;
    }

    return replacementId;
  }
}
