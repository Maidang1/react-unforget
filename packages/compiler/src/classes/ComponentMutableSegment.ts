import type * as babel from "@babel/core";
import * as t from "@babel/types";
import { makeDependencyCondition } from "~/ast-factories/make-dependency-condition";
import { hasHookCall } from "~/utils/is-hook-call";
import { Component } from "./Component";
import type { ComponentRunnableSegment } from "./ComponentRunnableSegment";
import type { ComponentVariable } from "./ComponentVariable";
import { DEFAULT_SEGMENT_CALLABLE_VARIABLE_NAME } from "~/utils/constants";
import { ComponentSegmentDependency } from "./ComponentSegmentDependency";

export const COMPONENT_MUTABLE_SEGMENT_COMPONENT_UNSET_TYPE = "Unset";
export const COMPONENT_MUTABLE_SEGMENT_COMPONENT_VARIABLE_TYPE =
  "ComponentVariable";
export const COMPONENT_MUTABLE_SEGMENT_COMPONENT_RUNNABLE_SEGMENT_TYPE =
  "ComponentRunnableSegment";

type ComponentMutableSegmentType =
  | typeof COMPONENT_MUTABLE_SEGMENT_COMPONENT_UNSET_TYPE
  | typeof COMPONENT_MUTABLE_SEGMENT_COMPONENT_VARIABLE_TYPE
  | typeof COMPONENT_MUTABLE_SEGMENT_COMPONENT_RUNNABLE_SEGMENT_TYPE;

export type SegmentTransformationResult = {
  prformTransformation: () => babel.NodePath<babel.types.Node>[] | null;
  segmentCallableId: babel.types.Identifier;
  dependencyConditions: babel.types.Expression | null;
  hasHookCall: boolean;
  hasReturnStatement?: boolean;
  updateCache?: babel.types.Statement | null;
  replacements: babel.types.Node[] | null;
} | null;

export abstract class ComponentMutableSegment {
  private segmentCallableId: t.Identifier | null = null;
  protected appliedTransformation = false;

  // Mutable code segments that this depends on this
  protected dependencies = new Set<ComponentSegmentDependency>();

  // Mutable code segments that are children of this
  protected children = new Set<ComponentMutableSegment>();

  protected parent: ComponentMutableSegment | null = null;

  constructor(
    public component: Component,
    parent: ComponentMutableSegment | null = null,
    protected type: ComponentMutableSegmentType = COMPONENT_MUTABLE_SEGMENT_COMPONENT_UNSET_TYPE
  ) {
    if (parent) {
      this.setParent(parent);
    }
  }

  setParent(parent: ComponentMutableSegment | null) {
    this.parent?.removeChild(this);
    this.parent = parent;
    parent?.addChild(this);
  }

  removeChild(child: ComponentMutableSegment) {
    this.children.delete(child);
  }

  addChild(child: ComponentMutableSegment) {
    this.children.add(child);
  }

  addDependency(componentVariable: ComponentVariable, accessorNode: t.Node) {
    if (this.isComponentVariable() && componentVariable === this) {
      return;
    }
    const componentSegmentDependency = new ComponentSegmentDependency(
      componentVariable,
      accessorNode as any
    );

    let alreadyHasDependency = false;
    for (const dependency of this.dependencies) {
      if (dependency.equals(componentSegmentDependency)) {
        alreadyHasDependency = true;
        break;
      }
    }

    if (!alreadyHasDependency) {
      this.dependencies.add(componentSegmentDependency);
    }
  }

  hasDependencies() {
    return this.dependencies.size > 0;
  }

  getDependencies() {
    return this.dependencies;
  }

  abstract computeDependencyGraph(): void;

  abstract get path(): babel.NodePath<babel.types.Node>;

  abstract applyTransformation(
    performReplacement?: boolean
  ): SegmentTransformationResult;

  protected makeDependencyCondition() {
    return makeDependencyCondition(this);
  }

  hasHookCall() {
    return hasHookCall(this.path, this.component.path);
  }

  getParentStatement() {
    const parentStatement = this.path.find(
      (p) => p.isStatement() && p.parentPath.isBlockStatement()
    ) as babel.NodePath<babel.types.Statement> | null;

    return parentStatement;
  }

  isComponentVariable(): this is ComponentVariable {
    return this.type === COMPONENT_MUTABLE_SEGMENT_COMPONENT_VARIABLE_TYPE;
  }

  isComponentRunnableSegment(): this is ComponentRunnableSegment {
    return (
      this.type === COMPONENT_MUTABLE_SEGMENT_COMPONENT_RUNNABLE_SEGMENT_TYPE
    );
  }

  protected makeSegmentCallStatement(
    transformation: SegmentTransformationResult
  ) {
    if (!transformation) {
      return null;
    }

    const {
      dependencyConditions,
      segmentCallableId,
      hasHookCall,
      updateCache,
      hasReturnStatement,
    } = transformation;

    const callSegmentCallable = t.callExpression(segmentCallableId, []);
    const updateStatements: t.Statement[] = [];

    if (hasReturnStatement) {
      const customCallVariable =
        this.component.path.scope.generateUidIdentifier();
      updateStatements.push(
        t.variableDeclaration("const", [
          t.variableDeclarator(customCallVariable, callSegmentCallable),
        ])
      );

      updateStatements.push(
        // if customCallVariable not equal to null, return it
        t.ifStatement(
          t.binaryExpression(
            "!==",
            customCallVariable,
            this.component.getCacheNullIdentifier()
          ),
          t.blockStatement([t.returnStatement(customCallVariable)])
        )
      );
    } else {
      updateStatements.push(t.expressionStatement(callSegmentCallable));
    }

    if (updateCache) {
      updateStatements.push(updateCache);
    }

    const callStatementWithCondition =
      dependencyConditions && !hasHookCall
        ? [
            t.ifStatement(
              dependencyConditions,
              t.blockStatement(updateStatements)
            ),
          ]
        : updateStatements;

    return callStatementWithCondition;
  }

  protected getSegmentCallableId() {
    if (!this.segmentCallableId) {
      this.segmentCallableId = this.component.path.scope.generateUidIdentifier(
        DEFAULT_SEGMENT_CALLABLE_VARIABLE_NAME
      );
    }

    return this.segmentCallableId;
  }
}
