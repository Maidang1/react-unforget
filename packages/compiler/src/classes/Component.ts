import type * as babel from "@babel/core";
import { Binding } from "@babel/traverse";
import * as t from "@babel/types";
import { isChildOfScope, isControlFlowStatement } from "~/utils/ast-tools";
import {
  DEFAULT_CACHE_COMMIT_VARIABLE_NAME,
  DEFAULT_CACHE_NULL_VARIABLE_NAME,
  DEFAULT_CACHE_VARIABLE_NAME,
  RUNTIME_MODULE_CREATE_CACHE_HOOK_NAME,
} from "~/utils/constants";
import { isInTheSameFunctionScope } from "~/utils/is-in-the-same-function-scope";
import { getFunctionParent } from "../utils/get-function-parent";
import { ComponentMutableSegment } from "./ComponentMutableSegment";
import { ComponentRunnableSegment } from "./ComponentRunnableSegment";
import { ComponentVariable } from "./ComponentVariable";

export class Component {
  private runnableSegments = new Map<
    babel.NodePath<babel.types.Statement>,
    ComponentRunnableSegment
  >();
  private componentVariables = new Map<Binding, ComponentVariable>();
  private cacheValueIdentifier: t.Identifier;
  private cacheCommitIdentifier: t.Identifier;
  private cacheNullIdentifier: t.Identifier;

  private rootSegment: ComponentRunnableSegment | null = null;

  private mapBlockStatementToComponentRunnableSegment = new Map<
    babel.NodePath<babel.types.BlockStatement>,
    ComponentRunnableSegment
  >();

  constructor(public path: babel.NodePath<babel.types.Function>) {
    path.assertFunction();

    this.cacheValueIdentifier = path.scope.generateUidIdentifier(
      DEFAULT_CACHE_VARIABLE_NAME
    );

    this.cacheCommitIdentifier = path.scope.generateUidIdentifier(
      DEFAULT_CACHE_COMMIT_VARIABLE_NAME
    );

    this.cacheNullIdentifier = path.scope.generateUidIdentifier(
      DEFAULT_CACHE_NULL_VARIABLE_NAME
    );
  }

  computeComponentSegments() {
    this.prepareComponentBody();

    const body = this.path.get("body");
    if (!body.isBlockStatement()) {
      return;
    }

    this.rootSegment = this.addRunnableSegment(body);
  }

  prepareComponentBody() {}

  hasComponentVariable(binding: Binding) {
    return this.componentVariables.has(binding);
  }

  getComponentVariable(binding: Binding) {
    return this.componentVariables.get(binding);
  }

  findPotentialParentForSegment(
    path: babel.NodePath<babel.types.Node>
  ): ComponentRunnableSegment | null {
    const blockOrControlFlowStatement = path.findParent(
      (innerPath) =>
        (innerPath.isBlockStatement() || isControlFlowStatement(innerPath)) &&
        innerPath.isDescendant(this.path) &&
        isInTheSameFunctionScope(innerPath, this.path)
    ) as babel.NodePath<babel.types.BlockStatement> | null;

    if (!blockOrControlFlowStatement) {
      return null;
    }

    const parent =
      this.mapBlockStatementToComponentRunnableSegment.get(
        blockOrControlFlowStatement
      ) ?? null;

    const ensuredParent = parent
      ? parent
      : this.addRunnableSegment(blockOrControlFlowStatement);

    return ensuredParent;
  }

  lock() {
    this.componentVariables.forEach((componentVariable) => {
      componentVariable.lock();
    });
    this.runnableSegments.forEach((runnableSegment) => {
      runnableSegment.lock();
    });
  }

  addComponentVariable(binding: Binding) {
    if (!this.isBindingInComponentScope(binding)) {
      return;
    }

    const { path } = binding;

    const parent = this.findPotentialParentForSegment(path);

    if (this.hasComponentVariable(binding)) {
      const componentVariable = this.getComponentVariable(binding)!;
      componentVariable.setParent(parent);
      return componentVariable;
    }

    // if (isForStatementInit(path)) {
    //   // This is for variables defined in the for statement
    //   // return;
    // }

    const componentVariable = new ComponentVariable(
      this,
      parent,
      binding,
      this.componentVariables.size
    );

    this.componentVariables.set(binding, componentVariable);

    componentVariable.unwrapAssignmentPatterns();
    componentVariable.computeDependencyGraph();

    this.statementsToMutableSegmentMapCache = null;
    return componentVariable;
  }

  addRunnableSegment(path: babel.NodePath<babel.types.Statement>) {
    if (path === this.rootSegment?.path) {
      return this.rootSegment;
    }

    if (!isInTheSameFunctionScope(path, this.path)) {
      const parent = this.findPotentialParentForSegment(path);
      return parent;
    }

    const parent = this.findPotentialParentForSegment(path);

    if (this.runnableSegments.has(path)) {
      const found = this.runnableSegments.get(path)!;

      if (parent !== found.getParent()) {
        found.getParent()?.removeChild(found);
        found.setParent(parent);
      }

      return found;
    }

    const runnableSegment = new ComponentRunnableSegment(this, parent, path);

    if (path.isBlockStatement()) {
      this.mapBlockStatementToComponentRunnableSegment.set(
        path,
        runnableSegment
      );
    }

    this.runnableSegments.set(path, runnableSegment);

    runnableSegment.computeDependencyGraph();
    this.statementsToMutableSegmentMapCache = null;

    return runnableSegment;
  }

  isTheFunctionParentOf(path: babel.NodePath<babel.types.Node>) {
    return getFunctionParent(path) === this.path;
  }

  getRootComponentVariables() {
    return [...this.componentVariables.values()].filter(
      (componentVariable) => !componentVariable.hasDependencies()
    );
  }

  getCacheVariableIdentifier() {
    return t.cloneNode(this.cacheValueIdentifier);
  }

  getCacheCommitIdentifier() {
    return t.cloneNode(this.cacheCommitIdentifier);
  }

  getCacheNullIdentifier() {
    return t.cloneNode(this.cacheNullIdentifier);
  }

  private statementsToMutableSegmentMapCache: Map<
    babel.NodePath<babel.types.Statement>,
    ComponentMutableSegment
  > | null = null;

  getStatementsToMutableSegmentMap() {
    if (this.statementsToMutableSegmentMapCache) {
      return this.statementsToMutableSegmentMapCache;
    }

    const statementsToMutableSegmentMap = new Map<
      babel.NodePath<babel.types.Statement>,
      ComponentMutableSegment
    >();

    const statementsMapSet = (segment: ComponentMutableSegment) => {
      const parent = segment.getParentStatement();
      if (parent && !parent.isBlockStatement()) {
        statementsToMutableSegmentMap.set(parent, segment);
      }
    };

    this.runnableSegments.forEach(statementsMapSet);
    this.componentVariables.forEach(statementsMapSet);

    this.statementsToMutableSegmentMapCache = statementsToMutableSegmentMap;

    return statementsToMutableSegmentMap;
  }

  applyTransformation() {
    if (!this.rootSegment) {
      throw new Error("Root segment not found");
    }

    this.lock();
    const cacheVariableDeclaration = this.makeCacheVariableDeclaration();

    const body = this.path.get("body");

    if (!body.isBlockStatement()) {
      return;
    }

    this.rootSegment.applyTransformation();

    body.unshiftContainer("body", cacheVariableDeclaration);
  }

  getFunctionBlockStatement(): babel.NodePath<babel.types.BlockStatement> | null {
    const path = this.path;
    if (path.isFunctionExpression()) {
      return path.get("body");
    }
    // Need to duplicate the check because the type guard with || is not working
    else if (path.isFunctionDeclaration()) {
      return path.get("body");
    } else if (path.isArrowFunctionExpression()) {
      const body = path.get("body");
      if (body.isBlockStatement()) {
        return body;
      }
    }

    return null;
  }

  private makeCacheVariableDeclaration() {
    const sizeNumber = t.numericLiteral(this.componentVariables.size);
    const declaration = t.variableDeclaration("const", [
      t.variableDeclarator(
        t.arrayPattern([
          this.cacheValueIdentifier,
          this.cacheCommitIdentifier,
          this.cacheNullIdentifier,
        ]),
        t.callExpression(t.identifier(RUNTIME_MODULE_CREATE_CACHE_HOOK_NAME), [
          sizeNumber,
        ])
      ),
    ]);

    t.addComment(
      sizeNumber,
      "leading",
      "\n" +
        Array.from(this.componentVariables.values())
          .map((componentVariable) => {
            return `${componentVariable.getIndex()} => ${
              componentVariable.name
            }`;
          })
          .join("\n") +
        "\n"
    );

    return declaration;
  }

  isBindingInComponentScope(binding: Binding) {
    return isChildOfScope(this.path.scope, binding.scope);
  }

  // --- DEBUGGING ---
  __debug_getComponentVariables() {
    return this.componentVariables;
  }

  __debug_dependencies() {
    return [
      ...[
        ...this.runnableSegments.values(),
        ...this.componentVariables.values(),
      ].map((segment) => {
        return (
          String(segment.path) +
          "depends on (" +
          [...segment.getDependencies().values()]
            .map((d) => d.componentVariable.name)
            .join(" , ") +
          ")"
        );
      }),
    ].join("\n");
  }
}
