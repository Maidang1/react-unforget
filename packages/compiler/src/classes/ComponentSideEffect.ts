import type * as babel from "@babel/core";
import * as t from "@babel/types";
import { Component } from "./Component";
import { ComponentMutableSegment } from "./ComponentMutableSegment";
import { getReferencedVariablesInside } from "~/utils/get-referenced-variables-inside";
import { convertStatementToSegmentCallable } from "~/ast-factories/convert-statement-to-segment-callable";

export class ComponentSideEffect extends ComponentMutableSegment {
  constructor(
    component: Component,
    private _path: babel.NodePath<babel.types.Statement>
  ) {
    super(component, "ComponentSideEffect");
  }

  computeDependencyGraph() {
    this.dependencies.clear();
    this.sideEffects.clear();

    const referencesInThisPath = getReferencedVariablesInside(this.path);

    referencesInThisPath.forEach((binding) => {
      const dependent = this.component.addComponentVariable(binding);

      if (dependent) {
        this.addDependency(dependent);
        dependent.addSideEffect(this);
      }
    });
  }

  applyTransformation(performReplacement = true) {
    if (this.appliedTransformation) {
      return null;
    }

    const hasHookCall = this.hasHookCall();

    const { segmentCallableId, newPaths, replacements } =
      convertStatementToSegmentCallable(this.path, { performReplacement });
    this.appliedTransformation = true;

    return {
      replacements,
      segmentCallableId,
      newPaths,
      hasHookCall,
      dependencyConditions: this.makeDependencyCondition(),
    };
  }

  get path() {
    return this._path;
  }
}
