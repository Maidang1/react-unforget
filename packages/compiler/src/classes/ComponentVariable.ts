import { Binding } from "@babel/traverse";
import * as t from "@babel/types";
import { getReferencedVariablesInside } from "~/utils/get-referenced-variables-inside";
import { getDeclaredIdentifiersInLVal } from "../utils/get-declared-identifiers-in-lval";
import { Component } from "./Component";
import { getRightmostIdName } from "~/utils/get-rightmost-id-name";
import { isHookCall } from "~/utils/is-hook-call";

export class ComponentVariable {
  // The side effects of this variable
  private sideEffects = new Set<babel.NodePath<babel.types.Node>>();

  // ComponentVariables that reference this
  private dependents = new Map<string, ComponentVariable>();

  // ComponentVariables that this depends on
  private dependencies = new Map<string, ComponentVariable>();

  constructor(
    public binding: Binding,
    public component: Component,
    private index: number
  ) {}

  get name() {
    return this.binding.identifier.name;
  }

  getIndex() {
    return this.index;
  }

  addSideEffect(path: babel.NodePath<babel.types.Node>) {
    this.sideEffects.add(path);
  }

  addDependency(componentVariable: ComponentVariable) {
    this.dependencies.set(componentVariable.name, componentVariable);
  }

  computeDependencyGraph() {
    this.binding.referencePaths.forEach((referencePath) => {
      const parentVariableDeclarator = referencePath.findParent(
        (p) =>
          p.isVariableDeclarator() && this.component.isTheFunctionParentOf(p)
      ) as babel.NodePath<babel.types.VariableDeclarator> | null;

      if (parentVariableDeclarator) {
        const lval = parentVariableDeclarator.get("id");

        const ids = getDeclaredIdentifiersInLVal(lval);

        ids.forEach((id) => {
          let dependent = this.component.getComponentVariable(id);

          if (!dependent) {
            const binding = this.component.path.scope.getBinding(id);

            if (binding) {
              dependent = this.component.addComponentVariable(binding);
            }
          }

          if (dependent) {
            this.dependents.set(id, dependent);
            dependent.addDependency(this);
          }
        });
      } else {
        // TODO: Side effect calculation
        // console.log("reference parent ->", referencePath.parentPath?.node);
      }
    });

    // console.log("this.binding.path.node", this.binding.path.node);

    const referencedVariablesInDeclaration = getReferencedVariablesInside(
      this.binding.path
    );

    referencedVariablesInDeclaration.forEach((binding) => {
      this.component.addComponentVariable(binding);
    });
  }

  isDerived() {
    return this.dependencies.size > 0;
  }

  isHook() {
    if (this.isDerived()) {
      return false;
    }

    const parentVariableDeclarator = this.binding.path.findParent(
      (p) => p.isVariableDeclarator() && this.component.isTheFunctionParentOf(p)
    ) as babel.NodePath<babel.types.VariableDeclarator> | null;

    if (!parentVariableDeclarator) {
      return false;
    }

    // return true if init is a call expression of hooks

    const init = parentVariableDeclarator.get("init");

    if (!init.isCallExpression()) {
      return false;
    }

    return isHookCall(init);
  }

  private getCacheAccessorExpression() {
    t.memberExpression(
      this.component.getCacheVariableIdentifier(),
      t.numericLiteral(this.index)
    );
  }

  // --- DEBUGGING ---
  __debug_getDependents() {
    return this.dependents;
  }

  __debug_getDependencies() {
    return this.dependencies;
  }

  __debug_getSideEffects() {
    return this.sideEffects;
  }
}
