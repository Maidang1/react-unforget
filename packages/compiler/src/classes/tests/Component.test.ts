import { Component } from "~/classes/Component";
import { parse } from "~/utils/testing";

const parseCodeAndRun = (code: string, key: string) => {
  const path = parse(code);
  return path.get(key) as babel.NodePath<babel.types.Function>;
};

const FIXTURE_1 = `
function MyComponent() {
  const [state, setState] = useState(0);

  const myDerivedVariable = state + 1;

  const unusedVariable = 1;

  return <div>{myDerivedVariable}</div>;
}
`;

const FIXTURE_2 = `
function MyComponent({someProp}) {
  const [state, setState] = useState(0);

  const myDerivedVariable = state + 1;

  const unusedVariable = 1;

  return <div>{myDerivedVariable} {someProp}</div>;
}
`;

describe("ComponentVariable", () => {
  describe("computeDependencyGraph", () => {
    it("basic example", () => {
      const fn = parseCodeAndRun(FIXTURE_1, "body.0");

      const component = new Component(fn);

      const componentVariables = component.__debug_getComponentVariables();

      expect(componentVariables.size).toStrictEqual(3);
      expect([...componentVariables.keys()]).toStrictEqual([
        "myDerivedVariable",
        "state",
        "setState",
      ]);

      // state has myDerivedVariable as dependent
      expect([
        ...componentVariables.get("state")!.__debug_getDependents().keys(),
      ]).toStrictEqual(["myDerivedVariable"]);

      // state depends on nothing
      expect([
        ...componentVariables.get("state")!.__debug_getDependencies().keys(),
      ]).toStrictEqual([]);

      // setState depends on nothing
      expect([
        ...componentVariables.get("setState")!.__debug_getDependencies().keys(),
      ]).toStrictEqual([]);

      // myDerivedVariable depends on state
      expect([
        ...componentVariables
          .get("myDerivedVariable")!
          .__debug_getDependencies()
          .keys(),
      ]).toStrictEqual(["state"]);

      // myDerivedVariable has no dependents
      expect([
        ...componentVariables
          .get("myDerivedVariable")!
          .__debug_getDependents()
          .keys(),
      ]).toStrictEqual([]);
    });
  });

  describe("getRootComponentVariables", () => {
    it("returns all root component variables", () => {
      const fn = parseCodeAndRun(FIXTURE_2, "body.0");

      const component = new Component(fn);

      const rootComponentVariables = component.getRootComponentVariables();

      expect(
        rootComponentVariables.map((v) => v.binding.identifier.name)
      ).toEqual(["state", "setState", "someProp"]);
    });
  });

  describe("applyModification", () => {
    it("adds a cache variable declaration", () => {
      const fn = parseCodeAndRun(FIXTURE_2, "body.0");

      const component = new Component(fn);

      component.applyModification();

      const code = component.path.toString();

      expect(code).toMatchSnapshot();
    });
  });
});
