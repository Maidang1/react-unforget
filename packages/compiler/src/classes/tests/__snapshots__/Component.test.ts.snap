// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComponentVariable applyModification adds a cache variable declaration 1`] = `
"function MyComponent({
  someProp
}) {
  /*
  0: myDerivedVariable
  1: state
  2: setState
  3: someProp
  */
  const _$unforget = useCreateCache$unforget(4);
  const [state, setState] = useState(0);
  const myDerivedVariable = state + 1;
  const unusedVariable = 1;
  return <div>{myDerivedVariable} {someProp}</div>;
}"
`;
